gui/mass-remove.lua

```diff
@@ -0,0 +1,336 @@
+--building/construction mass removal/suspension tool
+
+--[====[
+
+gui/mass-remove
+===============
+Allows removal of buildings/constructions and suspend/unsuspend using
+a box selection.
+]====]
+
+local gui = require "gui"
+local guidm = require "gui.dwarfmode"
+local persistTable = require 'persist-table'
+
+MassRemoveUI = defclass(MassRemoveUI, guidm.MenuOverlay)
+
+--[====[
+The actions that can be taken. Key bindings in parentheses.
+
+--suspend (s): suspends the construction of a planned building/construction
+--unsuspend (p): resumes the construction of a planned building/construction
+--remove_n (n): designates a construction (wall, floor, etc) for removal. Similar to the native Designate->Remove Construction menu in DF
+--unremove_n (c): cancels removal of a construction (wall, floor, etc)
+--remove_x (x): designates a building (door, workshop, etc) for removal. Similar to the native Set Building Tasks/Prefs->Remove Building menu in DF
+--unremove_x (b): cancels removal of a building (door, workshop, etc)
+--remove_a (a): designates both constructions and buildings for removal, and deletes planned buildings/constructions
+--unremove_a (u): cancels removal designations for both constructions and buildings
+]====]
+local actions={"suspend", "unsuspend", "remove_n", "unremove_n", "remove_x", "unremove_x", "remove_a", "unremove_a"}
+--used to iterate through actions with + and -
+local action_indexes={suspend=1,unsuspend=2,remove_n=3,unremove_n=4,remove_x=5,unremove_x=6,remove_a=7,unremove_a=8}
+local actions_length=tablelength(actions)
+
+MassRemoveUI.ATTRS {
+    action="remove_a",
+    marking=false,
+    mark=nil
+}
+
+--Helper functions.
+local function tablelength(T)
+  local count = 0
+  for _ in pairs(T) do count = count + 1 end
+  return count
+end
+
+--Needed because I couldn't find a direct way of getting a tile at certain coordinates.
```

Comment:
I think what you're looking for is dfhack.maps.getTileFlags()

See https://github.com/DFHack/scripts/blob/19f84c9a8f4acac933d65c8dfe5e396f4caf84bd/internal/quickfort/dig.lua#L345 where quickfort does something similar

Reply:
Huh, you can mutate the returned flag table, and that affects the tile itself. I assumed it would be a copy or something like that.

Thanks for the reference to quickfort, I picked up another trick from over there. Instead of searching for jobs that remove constructions, and canceling them, I can just set the dig flag to `No` and also set the block's `designated` flag to true, and the game will look for jobs associated with constructions that are not being removed anymore, and cancel them. Saves me a lot of work!

Implemented in cff1bde